codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,289 @@
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        if (pState.hasFrame) {
+            int nsent = min(maxBytes, pState.streamableBytes);
+            pState.write(nsent, writer);
+            connectionState.pseudoTime += nsent;
+            return nsent;
+        }
+
+        return distributeToChildren(maxBytes, writer, pState);
+    }
+
+    private int distributeToChildren(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        long oldTotalQueuedWeights = pState.totalQueuedWeights;
+        State iState = pState.poll();
+        State iNextState = pState.peek();
+        try {
+            assert iNextState == null || iNextState.pseudoTime >= iState.pseudoTime : "iNextState.pseudoTime(" +
+                    iNextState.pseudoTime + ") < " + " iState.pseudoTime(" + iState.pseudoTime + ")";
+            int nsent = distribute(iNextState == null ? maxBytes :
+                                    min(maxBytes, (int) ((iNextState.pseudoTime - iState.pseudoTime) *
+                                             ((double) iNextState.stream.weight() / oldTotalQueuedWeights)
+                                             + allocationQuantum)),
+                               writer,
+                               iState);
+            iState.updatePseudoTime(nsent, oldTotalQueuedWeights);
+            return nsent;
+        } finally { // Do in finally to ensure the internal state is not corrupted if an exception is thrown.
+            // The offer operation is delayed until we unroll up the recursive stack, so we don't have to remove from
+            // the priority queue due to a write operation.
+            if (iState.hasFrameCountForTree != 0) {
+                pState.offer(iState);
+            }
+        }
+    }
+
+    private State state(Http2Stream stream) {
+        return stream.getProperty(stateKey);
+    }
+
+    /**
+     * For testing only!
+     */
+    int streamableBytes(Http2Stream stream) {
+        return state(stream).streamableBytes;
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
+        private final Queue<State> queue;
+        int streamableBytes;
+        int hasFrameCountForTree;
+        long pseudoTime;
+        long totalQueuedWeights;
+        boolean hasFrame;
+        private boolean isQueued;
+
+        State(Http2Stream stream) {
+            this(stream, 2, connectionState.pseudoTime);
+        }
+
+        State(Http2Stream stream, int initialSize, long pseudoTime) {
+            this.stream = stream;
+            queue = new PriorityQueue<State>(initialSize, WFQ_COMPARATOR);
```

Comment:
Potentially however this may make potentially costly removal operations more common. For example if a node goes from "empty to non-empty", then we will have to remove all its children (or descendants) from the queue.  The "empty to non-empty" event could be "no data to data" (maybe less common?) or I think "stream window exhausted" to "window update gives bytes back to stream window" (maybe more common?).

Here is some cost estimation of the additional removals (with a priority heap):
- Locate descendants to remove - O(height of priority tree from the node + number of children for that node) if we add extra state to each node which indicates it has children that are currently queued.
- Iteration cost for removal from queue - O(number elements in queue) with current PriorityQueue implementation, O(1) with https://github.com/netty/netty/pull/4538#discussion_r46862034
- Maintain Heap Priority - O(log(number elements in priority queue)) - we will still have to maintain the priority heap properties.

WDYT?